lib/ash/data_layer/mnesia/transformers/define_records.ex

```diff
@@ -0,0 +1,258 @@
+  # doesn't work in macros very well though. This function is creating the AST
+  # for assigning records at runtime. So if you do a `resource.to_ex_record(%{b:
+  # "foo"})` it will fill in the missing fields correctly. This just adds
+  # defaults of `nil` because `Ash.Resource` is handling defaults, but we could
+  # extend this to handle other types of defaults as well.
+  defp build_field_assignments(field_names, is_compound_key, attributes, primary_key) do
+    if is_compound_key do
+      # First field is the compound key tuple
+      pk_tuple_assignment =
+        quote do
+          {
+            unquote_splicing(
+              Enum.with_index(primary_key, 1)
+              |> Enum.map(fn {pk, idx} ->
+                expand_value(pk, idx)
+              end)
+            )
+          }
+        end
+
+      # Non-PK field assignments
+      non_pk_fields = Enum.reject(attributes, &(&1.name in primary_key))
+
+      non_pk_assignments =
+        Enum.with_index(non_pk_fields, length(primary_key) + 1)
+        |> Enum.map(fn {attr, idx} ->
+          {attr.name, expand_value(attr.name, idx)}
+        end)
+
+      [{:_pkey, pk_tuple_assignment} | non_pk_assignments]
+    else
+      Enum.with_index(field_names, 1)
+      |> Enum.map(fn {field, idx} ->
+        # {field, quote(do: Map.get(attrs, unquote(field), default))}
+        {field, expand_value(field, idx)}
+      end)
+    end
+  end
+
+  defp expand_value(field, idx) do
+    quote do
+      if default == :position do
+        Map.get(attrs, unquote(field), :"$#{unquote(idx)}")
+      else
+        Map.get(attrs, unquote(field), default)
+      end
+    end
+  end
+
+  defp inject_record_code(dsl_state, table, fields, attributes, primary_key) do
+    is_compound_key = length(primary_key) > 1
+    field_names = Enum.map(attributes, & &1.name)
+
+    field_assignments =
+      build_field_assignments(field_names, is_compound_key, attributes, primary_key)
+
+    Transformer.eval(
+      dsl_state,
+      [
+        table: table,
+        fields: fields,
+        field_names: field_names,
+        primary_key: primary_key,
+        is_compound_key: is_compound_key,
+        field_assignments: field_assignments
+      ],
+      quote do
+        require Record
+
+        Record.defrecordp(unquote(table), unquote(fields))
+
+        # Store field defaults as module attribute
+        @is_compound_key unquote(is_compound_key)
+        @primary_key_fields unquote(primary_key)
+
+        @doc """
+        Returns the field names for the Mnesia record, suitable for use with
+        `Mnesia.create_table/2` as the `:attributes` option.
+
+        For compound primary keys, returns `[:_pkey | non_pk_fields]`.
+        For simple primary keys, returns all attribute names.
+
+        ## Examples
+
+            iex> #{inspect(__MODULE__)}.mnesia_record_info()
+            [:id, :name, :age, :email]
+        """
+        def mnesia_record_info do
```

Comment:
As opposed to defining functions inside the resource, we should use functions in `Ash.DataLayer.Mnesia.Info`, and use the transformer to put this information into the persisted data, i.e `Spark.Dsl.Transformer.persist`

Reply:
👍  - I'll see if I can figure this out. Are you talking about just this function or reorganizing them all this way? If so, I might need an example to point to.